File: Easy/118 - PascalsTriangle/Solution.cs
/// 118. Pascal's Triangle - 杨辉三角
/// Difficulty: Easy
/// 执行用时: 268 ms, 在所有 C# 提交中击败了 22.35% 的用户
/// 内存消耗: 25.9 MB, 在所有 C# 提交中击败了 7.69% 的用户

public class Solution
{
    public IList<IList<int>> Generate(int numRows)
    {
        IList<IList<int>> triangleRows = new List<IList<int>>();
        if (numRows == 0) { return triangleRows; }

        triangleRows.Add(new List<int> { 1 });
        if (numRows == 1) { return triangleRows; }

        triangleRows.Add(new List<int> { 1, 1 });
        if (numRows == 2) { return triangleRows; }

        for (int i = 2; i < numRows; i++)
        {
            List<int> newRow = new List<int> { 1 };
            List<int> rowAbove = new List<int>(triangleRows[i - 1]);

            for (int j = 1; j < i; j++)
            {
                newRow.Add(rowAbove[j - 1] + rowAbove[j]);
            }

            newRow.Add(1);
            triangleRows.Add(newRow);
        }

        return triangleRows;
    }
}

File: Easy/118 - PascalsTriangle/solution.ts
// 118. Pascal's Triangle - 杨辉三角

/**
 * 执行用时：76 ms, 在所有 TypeScript 提交中击败了 90.20% 的用户
 * 内存消耗：40.1 MB, 在所有 TypeScript 提交中击败了 56.00% 的用户
 *
 * @param {number} numRows
 * @returns {number[][]}
 */
function generate(numRows: number): number[][] {
  const pascalTriangle: number[][] = [];
  if (numRows === 0) {
    return pascalTriangle;
  }

  pascalTriangle.push([1]);
  if (numRows === 1) {
    return pascalTriangle;
  }

  pascalTriangle.push([1, 1]);
  if (numRows === 2) {
    return pascalTriangle;
  }

  for (let i = 3; i <= numRows; i++) {
    const currentRow: number[] = [1];
    const previousRow: number[] = pascalTriangle[i - 2];
    for (let j = 1; j < i - 1; j++) {
      currentRow.push(previousRow[j - 1] + previousRow[j]);
    }
    currentRow.push(1);
    pascalTriangle.push(currentRow);
  }

  return pascalTriangle;
}
